src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,655 @@
+            if (!typeString) {
+                return undefined;
+            }
+
+            const closeParenToken = getFirstChildOfKind(getAccessorDeclaration, sourceFile, SyntaxKind.CloseParenToken);
+            return createCodeActions(getAccessorDeclaration.name.getText(), closeParenToken.getEnd(), `: ${typeString}`);
+        }
+
+        function createCodeActions(name: string, start: number, typeString: string) {
+            return [{
+                description: formatStringFromArgs(getLocaleSpecificMessage(Diagnostics.Infer_type_of_0_from_usage), [name]),
+                changes: [{
+                    fileName: sourceFile.fileName,
+                    textChanges: [{
+                        span: { start, length: 0 },
+                        newText: typeString
+                    }]
+                }]
+            }];
+        }
+
+        function getReferences(token: PropertyName | Token<SyntaxKind.ConstructorKeyword>) {
+            const references = FindAllReferences.findReferencedSymbols(
+                program,
+                cancellationToken,
+                program.getSourceFiles(),
+                token.getSourceFile(),
+                token.getStart());
+
+            Debug.assert(!!references, "Found no references!");
+            Debug.assert(references.length === 1, "Found more references than expected");
+
+            return map(references[0].references, r => <Identifier>getTokenAtPosition(program.getSourceFile(r.fileName), r.textSpan.start, /*includeJsDocComment*/ false));
+        }
+
+        function inferTypeForVariableFromUsage(token: Identifier) {
+            return InferFromReference.inferTypeFromReferences(getReferences(token), checker, cancellationToken);
+        }
+
+        function inferTypeForParametersFromUsage(containingFunction: FunctionLikeDeclaration) {
+            switch (containingFunction.kind) {
+                case SyntaxKind.Constructor:
+                case SyntaxKind.FunctionExpression:
+                case SyntaxKind.FunctionDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                    const isConstructor = containingFunction.kind === SyntaxKind.Constructor;
+                    const searchToken = isConstructor ?
+                        <Token<SyntaxKind.ConstructorKeyword>>getFirstChildOfKind(containingFunction, sourceFile, SyntaxKind.ConstructorKeyword) :
+                        containingFunction.name;
+                    if (searchToken) {
+                        return InferFromReference.inferTypeForParametersFromReferences(getReferences(searchToken), containingFunction, checker, cancellationToken);
+                    }
+            }
+        }
+
+        function getTypeAccessiblityWriter() {
+            if (!writer) {
+                let str = "";
+                let typeIsAccessible = true;
+
+                const writeText: (text: string) => void = text => str += text;
+                writer = {

```

Comment:
this code looks copied from somewhere. is it too hard to export it?

Reply:
I need to move this to a change tracker and use typeToTypeNode, but waiting on Wesley's changes.